### PR TITLE
Support gs urls as inputs

### PR DESCRIPTION
**Pull request by mckinsel**
Also, support nested structures as inputs and input files with the same name.

### CODE REVIEW

File: external_executors/local_cromwell/local_cromwell.py [hunk 1]
```diff
@@ -103,16 +103,47 @@
...
+
+def localize_url(url):
+    tmp_dir = tempfile.mkdtemp()
+    if url.startswith("https://") or url.startswith("http://"):
+        local_path = os.path.join(tmp_dir, os.path.basename(url))
+        subprocess.run(["curl", "-o", local_path, url], check=True)
+        return local_path
+    elif url.startswith("gs://"):
+        client = storage.client.Client.create_anonymous_client()
+        path_parts = url[5:].split("/")
```
Comment (ambrosejcarr) [inline external_executors/local_cromwell/local_cromwell.py]: I like `url.replace("gs://", "")` for readability, but your version is fine too.

Comment (ambrosejcarr) [reply external_executors/local_cromwell/local_cromwell.py]: If this is python3, you could do: 
```
bucket_name, *blob_parts = url.replace("gs://", "").split("/")
blob = "/".join(blob_parts)
```

File: external_executors/local_cromwell/local_cromwell.py [hunk 2]
```diff
@@ -103,16 +103,47 @@
...
+        bucket = client.bucket(bucket_name)
+        blob = bucket.blob('/'.join(path_parts[1:]))
+        local_path = os.path.join(tmp_dir, path_parts[-1])
+        blob.download_to_filename(local_path)
+        return local_path
+    else:
+        return url
+
+def localize_inputs(obj, match_func, localize_func):
+    if isinstance(obj, list):
```
Comment (ambrosejcarr) [inline external_executors/local_cromwell/local_cromwell.py]: would make this `isinstance(obj, (list, tuple))`

**Review by ambrosejcarr — APPROVED** [review]
Two trivial and optional observations of personal preference, one suggestion regarding type-matching.

File: external_executors/local_cromwell/local_cromwell.py (hunk 1 above)
Comment (mckinsel) [reply external_executors/local_cromwell/local_cromwell.py]: This feels very fancy. I will change.

File: external_executors/local_cromwell/local_cromwell.py (hunk 2 above)
Comment (mckinsel) [reply external_executors/local_cromwell/local_cromwell.py]: I can change this, though I think we're only going to get types produces by `json.load`, which shouldn't make a `tuple`: https://github.com/python/cpython/blob/3.6/Lib/json/decoder.py#L258-L276

**Review by ambrosejcarr — COMMENTED** [review]
The changes here look good. I have a general question about the lambda function structure. It seems like the language flow is at some points quite complex: 

e.g. there are places where it is `python[2.7?, 3?] < bash < python3`

It looks like this could have originally been a bash script that got ported into python. Is there value in extracting the python now that it's running in lambda, or is that wasted work?

**Comment by mckinsel** [comment]
So these are the constraints:

- lambdas themselves have to be written python, node, java, go, or C#. Python is probably the best option here.
- "User data" has to be able to run at instance boot, and I've only ever seen it as a bash script. It's possible it could be python though.
- There were some steps that were a real hassle or impossible in bash, so I switched back to a python heredoc for those.

That's what led to the unfortunate python(bash(python)) structure. Sadly, it's even more complicated than that. The "user data" that gets run on the EC2 instance at boot is limited in size. The script that actually downloads the inputs and runs cromwell can be too big. So there are actually two scripts: one that is the real user data, and one the gets downloaded to the instance and executed by the user data.

So one option would be to port the user data script and the cromwell script to python and see if that works. That would result in lots of lines that are like

```python
subprocess.run(["yum", "install", "-y", "docker"], check=True)
```

which I tend to think is more readable in bash.

**Pull request merged**